src/Riok.Mapperly/Descriptors/UserMethodMappingExtractor.cs

```diff
@@ -264,27 +268,32 @@ bool isStatic
         // target parameter is the second parameter (except if the reference handler is the first or the second parameter)
         // if the method returns void, a target parameter is required
         // if the method doesnt return void, a target parameter is not allowed
-        var targetParameter = MethodParameter.Wrap(
-            method.Parameters.FirstOrDefault(p => p.Ordinal != sourceParameter.Value.Ordinal && p.Ordinal != refHandlerParameterOrdinal)
-        );
-        if (method.ReturnsVoid == !targetParameter.HasValue)
-        {
-            parameters = null;
-            return false;
-        }
-
-        if (targetParameter.HasValue)
+        MethodParameter? targetParameter = null;
+        if (method.ReturnsVoid)
         {
+            targetParameter = MethodParameter.Wrap(
+                method.Parameters.FirstOrDefault(p => p.Ordinal != sourceParameter.Value.Ordinal && p.Ordinal != refHandlerParameterOrdinal)
+            );
+            if (!targetParameter.HasValue)
+            {
+                parameters = null;
+                return false;
+            }
             expectedParameterCount++;
         }
 
-        if (method.Parameters.Length != expectedParameterCount)
-        {
-            parameters = null;
-            return false;
-        }
-
-        parameters = new MappingMethodParameters(sourceParameter.Value, targetParameter, refHandlerParameter);
+        // if (method.Parameters.Length != expectedParameterCount)
+        // {
+        //     parameters = null;
+        //     return false;
+        // }
+        var extraParams = method.Parameters
+            .Skip(expectedParameterCount)
```

Comment:
Instead of skipping the expected parameter count, keep a set of already processed parameter ordinals and check the additional parameter ordinals are not included in this set.